### ISSUE TITLE
with-contract doesn’t properly adjust negative party for contracts broken within region

### ISSUE THREAD

**Issue by lexi-lambda** [racket/racket#2194]
The following program blames the wrong party:

```racket
(define/contract (f x)
  (-> integer? integer?)
  (add1 x))

(with-contract neg ()
  (f "hello"))
```

According to the documentation for `with-contract`, the party blamed should be `neg`, but instead the whole module is blamed. This happens whether `f` is defined with `define/contract` or defined in another module and assigned a contract using `contract-out`.

The same issue occurs when using `define/contract` instead of `with-contract`, presumably since it becomes `with-contract` internally:

```racket
(define/contract (f x)
  (-> integer? integer?)
  (add1 x))

(define/contract (g)
  (-> any)
  (f "hello"))
(g)
```

The above program should blame `g`, but instead it blames the whole module.

**Comment by mfelleisen** [comment]
This is a documentation bug. define/contract and with-contract create virtual submodules in the sense of establishing contract boundaries. So define/contract for f draws a contract boundary between (add1 x) and the module as a whole. Then with-contract neg draws another one. The surrounding module is the linking module, which receives f and then sends f into the virtual submodule without protecting it further. So it is to blame because it serves as a conduit without protection.

**Comment by rfindler** [comment]
Yes, this is the same conclusion I came to this morning too.

**Comment by rfindler** [comment]
(Well, with the additional thought that this is yet again an example of why `define/contract` should be avoided. It is *confusing*.)

**Comment by lexi-lambda** [comment]
I don’t understand this interpretation, and after talking to @chrdimo about it, he doesn’t either.

First, let me state what I think your interpretation *is*. I think what you’re saying is that writing

```racket
(define/contract (f x)
  (-> integer? integer?)
  (add1 x))
```

is equivalent to writing

```racket
(module f racket
  (provide (contract-out [f (-> integer? integer?)]))
  (define (f x) (add1 x)))
(require (rename-in 'f [f f*]))
(define f f*)
```

where the surrounding/requiring module serves as the “linking module” and therefore always gets blamed for any negative violations of `f`’s contract. But notice the dance I had to do with `rename-in` followed by `define` to get the behavior you describe, since a naïve translation to submodules doesn’t have that behavior without it:

```racket
#lang racket

(module f racket
  (provide (contract-out [f (-> integer? integer?)]))
  (define (f x) (add1 x)))

(require 'f)
(provide f)

(module* neg racket
  (require (submod ".."))
  (f "hello"))
```

Running the `neg` submodule above produces this contract error:

```
f: contract violation
  expected: integer?
  given: "hello"
  in: the 1st argument of
      (-> integer? integer?)
  contract from: (anonymous-module f)
  blaming: (anonymous-module neg)
   (assuming the contract is correct)
```

…which you’ll note blames `neg`, not the enclosing module (`anonymous-module` in this case).

Put another way, you seem to claim that the “linking module” (whatever that means precisely) should take the blame for all negative contract violations, since there is no explicit contract between the consumer and the linking module. But if this is true, then `contract-out` is broken, since reprovided identifiers with contracts on them blame the module they’re eventually used in, not the module that reprovided them.

Based on that reasoning, it seems to me that either `with-contract` is broken or `contract-out` is, and I’d have a hard time seeing how both could be right.

**Comment by rfindler** [comment]
Short answer for @chrdimo (and maybe useful for you too, Alexis) : there are three parties in play; the original `define/contract`. the `with-contract`, and the module that contains all of them. The agreement between the containing module and the `with-contract` is `any/c` and the agreement between the `define/contract` and the enclosing module is the contract written in the `define/contract`. There is no direct connection between `with-contract` and the `define/contract`.

**Comment by lexi-lambda** [comment]
I understand that point of view (and it makes sense to me in isolation), but the point of my previous comment is that it is inconsistent with the behavior of `contract-out`. Maybe this program will be more illustrative:

```racket
#lang racket

(module a racket
  (provide (contract-out [f (-> integer? integer?)]))
  (define (f x) (add1 x)))

(module b racket
  (require (submod ".." a))
  (provide f))

(module c racket
  (require (submod ".." b))
  (f "hello"))

(require 'c)
```
```
f: contract violation
  expected: integer?
  given: "hello"
  in: the 1st argument of
      (-> integer? integer?)
  contract from: (anonymous-module a)
  blaming: (anonymous-module c)
   (assuming the contract is correct)
```

As far as I can tell, exactly the same logic applies to this program. There are three parties in play, modules `a`, `b`, and `c`. The agreement between `c` and `b` is `any/c`, and the agreement between `b` and `c` is the contract written in the use of `contract-out`. However, the contract error blames `c`, not `b`.

I would be satisfied with either semantics for assigning blame—one does not seem meaningfully better to me than the other—but I would expect consistency throughout the contract system about which one is used.

**Comment by rfindler** [comment]
If I am not getting confused, this is a bug in either the contract system or a shortcoming in racket's provides that does not let us implement what we want to implement. If you actually write `contract-out` in `b` and put `any/c`, you'll see the right behavior.

**Comment by rfindler** [comment]
You would know better than me at this point!!

(Although contract-out is doing the right thing, right?)

**Comment by samth** [comment]
I think changing `provide` so that re-`provide` produced a different identifier would break a lot of assumptions that other Racket programs make.

**Comment by rfindler** [comment]
Maybe there can be a way to let the macro that's provided know if a
reprovide happened?

**Comment by lexi-lambda** [comment]
After a little bit more thought, I do sort of question the usefulness of this model. If module A defines a function and puts a contract on it, and module B re-exports that function, then module C misuses the function, I would imagine that 95% of the time, module C is the module actually at fault. I do understand that the “correct” thing to do on the part of the programmer is for module B to re-export the function with the same contract (or, as the case may be, a different contract that module B expects the function to satisfy, so if the function changes, a positive contract violation is raised that blames B), but even if we could totally eliminate the unnecessary overhead of two identical contract wrappers, this seems unlikely to be consistently done in practice.

Put another way, if the model is our source of truth, and it implies what you say it does, then the existing implementation is wrong, but the incorrect implementation also seems more useful in a world in which not all exports have contracts (and reprovides essentially never have contracts) than a correct one would be. Is that not a little troubling?

(To be clear, I’m not suggesting that the model is wrong or even that the existing behavior is desirable, since its usefulness seems largely incidental rather than intentional, but I’m merely wondering aloud if there’s anything we could do to better serve both situations. I also realize an answer to that question is probably bigger than belongs in a GitHub issue! Given that, perhaps I should just close this bug in the meantime.)

**Comment by rfindler** [comment]
I can not argue with that.

All along we have been saying that "(provide x)" should be the same as
"(provide (contract-out [x any/c])". But maybe we should have been saying
not "any/c" in there but instead "whatever contract it had when it was
imported (which is any/c if contract-out wasn't used)".

Of course this still leaves the problem of how to implement it.

Robby

**Comment by lexi-lambda** [comment]
From a lunch conversation: the blame of `define/contract` is confusing to reason about because the regions are not intuitive and the boundary is ill-defined. Therefore, we might consider making `define/contract` into `define/assertion` that behaves like `invariant-assertion` and dispenses with blame.

**Comment by rfindler** [comment]
Slightly different words for saying the same thing: many people find define/contract confusing `because`, although it sets up this wonderful boundary around itself, that boundary is strangely permeable. If you are the `f` that got defined, well, values go thru the contract and things are fine, but if your function has a free variable in it then the boundary is totally confusing. Vincent put it nicely: it is a high-security safe .... with a cat door in it.

So, if we changed `define/contract` so that it was really an assertion definition mechanism then we would just leave out the blame information in the error messages and we would not have this problem anymore. No one would be confused by why it says that this thing or that thing was to blame, instead they would just see "I wanted an int but got a list".

Put yet another way: a strangely defined boundary (and this one is, due to free variables), while semantically sensible, is pragmatically terrible.

**Comment by rmculpepper** [comment]
FWIW, in mzscheme 4.2, the original example raises the following error: `(region neg) broke the contract (-> integer? integer?) on f; expected <integer?>, given: "hello"`, so I suspect bit rot.

The idea that requiring a module means accepting its contracts has always seemed like a bad model to me, mainly for the reasons that @lexi-lambda articulated. In my mind, a contract on a binding is accepted when a reference to that binding occurs in expression position. This is The Right Thing for provide, IMO, but The Wrong Thing for latent references in macro templates. Alas.

**Comment by mfelleisen** [comment]
I disagree with "bit rot." I think that 4.2 was in error, and that we fixed it. Can someone with more git fu please dig out the relevant commit message? 

When you accept the services of your landlord in Prague, you have no choice but to accept the legal context and (perhaps with small modifications) the contract that  your landlord asks you to sign.  Whether you move in or not, doesn't matter to this signing. In the same sense, if a module requires and re-provides services, it is a part of the chain of responsibility because the ultimate consumer may not even know about the original producer. [I could also consider the re-export position an expression position.]

**Comment by rmculpepper** [comment]
The last time I found an apartment in Boston, I used a realtor, but they were not a party to the lease contract. I'm sure they had a side contract with one of us to make sure they got their fee, which muddles the analogy, but I'm also sure they didn't want to hear about it if I didn't pay rent or if the landlord refused to fix the water heater. They just wanted to connect landlords and tenents. Similarly, if I buy a copy of Windows at Best Buy and make a thousand copies, it's not Best Buy that I'll be seeing in court.

The chain-of-responsibility argument leads to bad behavior by default, because (in this case) it reports the first party to not fill out the extra paperwork rather than the actual site of the misuse. That is, the error `the top level broke the contract...` is *worse* than `(region neg) broke the contract...`. At the module level, it would force intermediate modules (aggregators like `racket`) to mumble `recontract-out` over their exports or write the contracts out again (and as Sam pointed out, it would create lots of import conflicts). Can you modify the original example to express that the top level (or enclosing module) propagates the contract on `f` to any `with-contract` subexpressions that may appear within it, so that the chain-of-responsibility model blames the `neg` region?

**Comment by mfelleisen** [comment]
Wrong analogy. A realtor is much more like a distributed (and parallel) system. He has nothing to do with the seller. He doesn’t ever require the service. He knows about it and points you there. 

A much better analogy is that of a contractor and a sub-contracor, like the guy next door who’s a roofer but for building decks on flat roofs subcontracts to a guy who doesn’t even speak English. When things go wrong, the neighbors will first go after the contractor because they don’t even know about the sub. (They found out at the end when I asked for a bid from the sub directly by accident and the bid was half.) 

A module that requires and exports services is pretty much a contractor that delegates to a sub-contractor.

**Comment by rmculpepper** [comment]
I agree that the subcontractor analogy applies sometimes (that's why `recontract-out` exists), but treating it as the *only* analogy leads to a worse PL design. (Specifically, worse than "a contract is accepted by reference in expression position, not by require".) I think the analogies have become distracting; they should be sources of inspiration, not tools of analysis. We should talk about which model is better for Racket programmers. That's what I was trying to do in the second half of my last message.

**Comment by rfindler** [comment]
@rmculpepper I'm still mulling over this thing you're talking about wrt to macros and contracts. Can you elaborate a little bit? Are you saying that this program should result in blame being assigned to `n`?

```
#lang racket
(module m racket
  (provide
   (contract-out
    [f (-> integer? integer?)]))
  (define (f x) (+ (* x x x) (* x x) x)))

(module n racket
  (require (submod ".." m) syntax/parse/define)
  (provide m)
  (define-simple-macro
    (m e)
    (if e (f #f) (f 0))))

(module o racket
  (require (submod ".." n))
  (m #t))

(require 'o)
```

**Comment by rmculpepper** [comment]
@rfindler Yes, I think it would be reasonable to blame `n` with the rationale that the reference to `f` occurs in `n` as a latent reference in what turns out to be an expression position.

On the other hand, consider this example:

    (module m racket
      (provide (contract-out [f (-> integer? integer?)]))
      (define (f x) (+ (* x x x) (* x x) x)))
    (module p racket
        (require 'm) 
        (define-syntax f-info (list #'f))
        (provide f-info))
    (module q racket
        (require 'p)
        (define-syntax (use-f stx) #`(#,(car (syntax-local-value #'f-info)) #f))
        (use-f))
    (require 'q)

In that example I think the blame story is unclear, because the `#'f` identifier didn't really become a reference until `q` decided to use it that way. If `use-f` had expanded into, for example, `(provide f)` instead, I would not consider either `p` or `q` to have accepted the contract on `f`.

On the other hand, is there really a difference between this example and your macro example, or between this and a similar example with match expanders, which are like macros but not quite exactly? I'm not sure it's possible to find bright-line distinctions in a macro system like Racket's that delays commitment to the interpretation of an identifier so late. (Similar issues prompted the switch from the old certificate system to the current opt-in taint system for "macro" integrity.)

**Comment by mfelleisen** [comment]
When I run this (note correct syntax), I get what I think is correct blame in a language with compile-time manipulations. 

```
#lang racket
(module m racket
  (provide (contract-out [f (-> integer? integer?)]))
  (define (f x) (+ (* x x x) (* x x) x)))
(module p racket
  (require (submod ".." m))
  (define-syntax f-info (list #'f))
  (provide f-info))
(module q racket
  (require (submod ".." p))
  (define-syntax (use-f stx) #`(#,(car (syntax-local-value #'f-info)) #f))
  (use-f))
(require 'q)
```

I do wonder (1) whether @chrdimo 's "blame completeness" provides guidance here and (2) how his blame effectiveness work will pay off here. (Can't wait for @lexi-lambda to get some results on that work.)

**Comment by mfelleisen** [comment]
On the very original issue (re-providing and blame), I propose the following idea for guidance: 

- if you have modules A and B 
- and if you wish to provide services to someone w/o revealing that B (and its full exports) exists 
- then A should re-provide the necessary services and take the blame so that clients don't find out about B.

**Comment by rfindler** [comment]
@mfelleisen yes, I think that was the point of @rmculpepper 's comment. The other example, however, I think you could argue was wrong. And in the other example a fix would be to use the source location of the identifier as the negative party (which would break the example that you are saying works right).

**Comment by mfelleisen** [comment]
@rfindler I don't get why the two examples don't give us the correct blame story -- both. I disagree with @rmculpepper that "expression position" or "reference" is relevant.

**Comment by rfindler** [comment]
As for the "very original issue" comment, I think that the situation you describe is a common one, where B is actually named something like `A/private/one-of-my-many-files-that-I-broke-things-up-into` and `A` reprovides lots of things. In that case, it is inconvenient to get the right kind of blame for positive failures, as you'd like to avoid the mention of the private identifiers.

Generally what I've done in these situations (which is okay in some situations but not in others) is to put the contracts entirely in `A` and have `B` just `provide` them. 

But in any case, are we agreeing that there is a reasonable interpretation for any of these options, the issue is just what reprovide should be a short-hand for, in order to get what people want, most of the time? Both analogies (realtor & subcontract) can be made to be sensible at the lower, fully-explicated level?

**Comment by rfindler** [comment]
In my macro example, I think the blame should lie with `n`, but it lies with `o` when you run the program currently.

**Comment by mfelleisen** [comment]
@rfindler 

(1) I ran all examples and I find the current blame assignments correct based on my mental model (correctness and completeness). 

(2) I do not want to make judgments based on "reasonable interpretation." We have developed solid mathematical tools for interpretations (correctness, completeness) and we are developing pragmatic tools for evaluating the effectiveness of blame. If the latter differs from the former, I am willing to ditch the math if I understand the pragmatic results. 

I also think that `private` modules should come with contracts. If we say "don't use contracts on privates" our contract system is broken.

**Comment by rfindler** [comment]
Please explain how your mental model (or any calculus) accounts for macros. Did I miss a paper?

Re private, sure. I don't get how that comment is at all relevant to what I wrote.

**Comment by mfelleisen** [comment]
re: private ~~ You wrote "don't write contracts for `private` modules". I don't understand why we should suggest or impose such a rule. The interaction between `public` and `private` can be as complex as between `public` and `public`. So why not allow contracts? And why not allow re-exports so that we "hide" the existence of `private`.  If we do it is natural to blame the `public` re-exporter of the service. 

re: macros ~~ We have a bunch of simple rules. (1) Don't report a contract violation until we have a witness. (2) We look at the running code and make sure that all channels of communications are controllable. (I think this remains relatively intact in the presence of macros.) (3) We say correctness is when changing code in the blamed macro can extend the next run beyond the contract violation (but may of course raise another one after one step). 

So in example 2: Q is required, therefore the call to f is generated within Q and run, and therefore Q is blamed. In example 1: O is required, the call to f is generated within O and run, therefore O is blamed.

**Comment by rfindler** [comment]
re: private. That's not what wrote/meant. Of course, contracts should go everywhere that the programmer deems necessary. And my point is that when I go to the trouble to put contracts on the private ones and those are the final contracts, with the current "ergonomics" of the contract library, I have to duplicate the contracts (in order to the blame that makes sense to the client). That seems like poor design.

re: macros. in my mind, the sensible way to look at this is based on the "blame the file that needs to change to eliminate/shift the blame" and in that lens you clearly want to change `n`. Indeed, if the `define-simple-macro` were `define`, blame would fall on `n`. I don't see why the `define-simple-macro` version is different (from this particular perspective, of course -- macros are different in general, as @rmculpepper points out so "simple" fixes also don't seem right here).

**Comment by mfelleisen** [comment]
@rfindler 

re: private ~~ I don't get it. If I re-export, the contract system blames me, which is what I want. 

re: macros ~~ But the code  for the call shows up in O. If I don't use m, I don't get the violation. Correctness as we defined it. Correctness does *not* say this is the best possible change. 

re: macros ~~ Changing the `define-simple-macro` to `define` means the code for the call is inside of N and N must be blamed. 

More generally, as Ryan was wrapping up his PhD I said "we have made no progress in understanding macros separately from their expansion." I think this is what is happening here.

**Comment by rfindler** [comment]
re: private. From where I sit, you want something strange. I don't know what to say to you. I think it is a much more reasonable want to want something where the outer, non-private module gets blamed and for that to be convenient (ie not require you to duplicate the contracts (or roll your own contract-exporting dsl)). Do you really document your private modules? Does it really make sense for clients of your library to see their names in their error messages? Just to be clear, as I wrote before, re-exporting meaning "re-export with `any/c`" and re-exporting meaning "re-export with the same contract I got it as" are both sensible in different situations. Too bad only one of them is supported by the current version of `racket/contract`.

re macros: "the code is inside" isn't a technical term and I think reasonable people might disagree on your definition here. re "more generally": yes, indeed!

**Comment by mfelleisen** [comment]
re: private. We chose a version that I think works well. I write the contract where the function is defined, and re-exporting it from the public module gets the public one blamed. I don't get what we're arguing about. 

re: macros: "code inside" has a precise definition. Expand all macros and deal with the generated modules according to our model. Because of "in general".

**Comment by rfindler** [comment]
re: private. We chose a version that works well for some situations and works poorly for others. I think `racket/contract` lacks a way to express another commonly desired pattern (namely, export this with the same contract I got it as without making me write the contract down a second time). I don't get why this is a hard point to agree with.

re: macros: that's one definition. Another one might "I wrote it in this file" (as reported by the source location information when I query the syntax object). I don't see why yours is a priori better (really, neither is better in my mind, which is much of why I agree with your quote of Ryan).

**Comment by rmculpepper** [comment]
@rfindler Does `recontract-out` not cover what you mean by "export this with the same contract I got it as without making me write the contract down a second time"?

**Comment by rfindler** [comment]
Oh, almost! I forgot about that one. We need a `all-from-out` version.

**Comment by rmculpepper** [comment]
Back to the original issue: here's a slight variation on @lexi-lambda's original example:

    #lang racket
    (with-contract A
      ([f (-> integer? integer?)])
      (define (f x) (add1 x)))
    (with-contract B
      ()
      (f "hello"))

If I understand @mfelleisen correctly, the "subcontractor" model says that the contract on `f` is immediately accepted by the enclosing module before reaching the occurrence within the `B` region. So the contract violation blames the anonymous outer module.

The "contract accepted at the site of a reference" model says that the contract on `f` is accepted by the reference within `B`, so `B` should get blamed.

I find the second model both more useful and more intuitive. It is more useful because `B` more precisely pinpoints the place in the program that needs fixing. It is more intuitive because there is actually a syntactic anchor for the negative blame party (the `f` reference), whereas in the other model it seems like the contract is accepted by the empty space between the two `with-contract` forms.

**Comment by rmculpepper** [comment]
re: macros. If I understand Matthias's definition of correctness, then it would be correct to blame *either* N or O in the first macro example and *either* P or Q in the second, because changing either the definition or use site could eliminate the contract violation. But that should also be true if macros are not involved:

    #lang racket
    (module m racket
      (provide
       (contract-out
        [f (-> integer? integer?)]))
      (define (f x) (+ (* x x x) (* x x) x)))
    (module n racket
      (require (submod ".." m))
      (provide g)
      (define (g x) (if x (f #f) (f 0))))
    (module o racket
      (require (submod ".." n))
      (g #t))
    (require 'o)

Is O a correct candidate for blame in this example? If so, then what is the additional factor that makes us say that N is the better party to blame? Why does that reasoning not apply to the macro example as well?

The main meta-design trick we have for macros is to try to apply the principles that govern the behavior of functions to macros ... and see how far we get. Functions are lexically scoped; applying lexical scoping to macros yields hygiene. Functions do not expose their implementations; applying that to macros yields integrity protection (eg, syntax taints). It seems reasonable to at least try to make blame for macros work similarly to functions. (And based on previous experience it should not be surprising if murky corner cases remain.)
